File: src/main/java/random/telegramhomebot/auth/services/AppUserDetailsService.java
package random.telegramhomebot.auth.services;

import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;
import random.telegramhomebot.auth.AuthErrorCodes;
import random.telegramhomebot.auth.UserPrincipal;
import random.telegramhomebot.auth.entities.User;
import random.telegramhomebot.auth.repositories.UserRepository;
import random.telegramhomebot.services.MessageService;
import random.telegramhomebot.utils.NetUtils;

import javax.servlet.http.HttpServletRequest;

@RequiredArgsConstructor
@Service
public class AppUserDetailsService implements UserDetailsService {

    private final UserRepository userRepository;
    private final LoginAttemptService loginAttemptService;
    private final HttpServletRequest request;
    private final MessageService messageService;

    @Value("${login.blocking.time.in.minutes}")
    private int blockingTimeInMinutes;

    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        String ip = NetUtils.getClientIp(request);
        if (loginAttemptService.isBlocked(ip)) {
            throw new RuntimeException(messageService.getMessage(AuthErrorCodes.USER_BLOCKED.getErrorMessageCode(),
                    new Object[]{blockingTimeInMinutes / 60}));
        }

        User user = userRepository.findByUsername(username);
        if (user == null) {
            throw new UsernameNotFoundException("Cannot find username: " + username);
        }
        return new UserPrincipal(user);
    }
}

File: src/main/java/random/telegramhomebot/auth/services/AppUserDetailsService.kt
package random.telegramhomebot.auth.services

import org.springframework.beans.factory.annotation.Value
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service
import random.telegramhomebot.auth.UserPrincipal
import random.telegramhomebot.auth.db.repositories.UserRepository
import random.telegramhomebot.auth.enums.AuthErrorCode
import random.telegramhomebot.services.messages.MessageService
import random.telegramhomebot.utils.NetUtils.Companion.getClientIp
import javax.servlet.http.HttpServletRequest

@Service
class AppUserDetailsService(
    private val userRepository: UserRepository,
    private val loginAttemptService: LoginAttemptService,
    private val request: HttpServletRequest,
    private val messageService: MessageService
) : UserDetailsService {

    @Value("\${login.blocking.time.in.minutes}")
    private val blockingTimeInMinutes = 0

    @Throws(UsernameNotFoundException::class)
    override fun loadUserByUsername(username: String): UserDetails {
        val ip = getClientIp(request)
        if (loginAttemptService.isBlocked(ip)) {
            throw RuntimeException(
                messageService.getMessage(AuthErrorCode.USER_BLOCKED.messageCode, arrayOf(blockingTimeInMinutes / 60))
            )
        }
        val user = userRepository.findByUsername(username)
            ?: throw UsernameNotFoundException("Cannot find username: $username")
        return UserPrincipal(user)
    }
}
